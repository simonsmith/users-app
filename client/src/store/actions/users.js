import {makeApiRequest} from '../middleware/api';

const URL = 'http://localhost:5000/users';

export function requestAllUsers() {
  return makeApiRequest({
    url: URL,
    type: 'USERS',
  });
}
