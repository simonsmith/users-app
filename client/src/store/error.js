import {createReducer} from 'redux-starter-kit';

const initialState = {
  message: null,
  request: null,
};

export const errorReducer = createReducer(initialState, {
  API_REQUEST_FAILURE: (state, action) => {
    return {
      message: action.payload.toString(),
      response: action.payload.response,
    };
  },
});
