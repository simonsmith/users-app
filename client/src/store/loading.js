import {createReducer} from 'redux-starter-kit';

const initialState = {};

export const loadingReducer = createReducer(initialState, {
  API_REQUEST_START: (state, action) => {
    return {
      ...state,
      [action.payload]: true,
    };
  },
  API_REQUEST_END: (state, action) => {
    return {
      ...state,
      [action.payload]: false,
    };
  },
});
