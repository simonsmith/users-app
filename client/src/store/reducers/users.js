import {createReducer} from 'redux-starter-kit';

const initialState = {
  isLoading: false,
  result: [],
  error: null,
};

export default createReducer(initialState, {
  USERS_REQUEST: state => {
    state.isLoading = true;
  },
  USERS_GET_SUCCESS: (state, action) => {
    state.isLoading = false;
    state.result = action.payload.result;
  },
  USERS_FAILURE: (state, action) => {
    state.isLoading = false;
    if (action.error) {
      state.error = action.payload;
    }
  },
});
