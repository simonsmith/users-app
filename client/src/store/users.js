import {createReducer} from 'redux-starter-kit';
import {normalize, schema} from 'normalizr';
import {apiRequest} from './middleware/api';

export const FETCH_USERS = 'FETCH_USERS';
export const CREATE_USER = 'CREATE_USER';
export const FETCH_USERS_SUCCESS = 'FETCH_USERS_SUCCESS';

const role = new schema.Entity('role');
const user = new schema.Entity('users', {
  role,
});

function normalizeUsers({data}) {
  return {
    type: FETCH_USERS_SUCCESS,
    payload: normalize(data, [user]),
  };
}

export function createNewUser(user) {
  return apiRequest('/users', {
    type: CREATE_USER,
    onSuccess: requestAllUsers,
    method: 'POST',
    data: user,
  });
}

export function requestAllUsers() {
  return apiRequest('/users', {
    type: FETCH_USERS,
    onSuccess: normalizeUsers,
  });
}

const initialState = {
  result: [],
};

export const usersReducer = createReducer(initialState, {
  [FETCH_USERS_SUCCESS]: (state, action) => {
    state.result = action.payload.result;
  },
});

export function selectUsers(state) {
  return {
    result: state.users.result,
    entities: state.entities.users,
  };
}

export function selectRoles(state) {
  return {
    entities: state.entities.role,
  };
}
