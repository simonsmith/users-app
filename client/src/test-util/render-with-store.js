import React from 'react';
import {render} from '@testing-library/react';
import {Provider} from 'react-redux';
import createStore from '../store/create';

export function renderWithStore(component, store = createStore()) {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
}
