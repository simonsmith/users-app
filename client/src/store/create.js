import {configureStore} from 'redux-starter-kit';
import {createApiMiddleware} from '@simonsmith/redux-api-middleware';
import axios from 'axios';
import {usersReducer as users} from './users';
import {entitiesReducer as entities} from './entities';
import {loadingReducer as loading} from './loading';
import {errorReducer as error} from './error';

const BASE_URL = process.env.API_BASE_URL || 'http://localhost:5000/';

export default function() {
  return configureStore({
    reducer: {
      entities,
      users,
      loading,
      error,
    },
    middleware: [
      createApiMiddleware(axios, {
        requestDefaults: {
          baseURL: BASE_URL,
        },
      }),
    ],
  });
}
