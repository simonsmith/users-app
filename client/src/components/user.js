import React, {useState} from 'react';
import {updateSingleUser} from '../store/users';
import {UserEditForm} from './user-edit-form';

export function User(props) {
  const [isEditing, setEditing] = useState(false);
  return (
    <div>
      <button
        onClick={() => {
          setEditing(!isEditing);
        }}
      >
        {isEditing ? 'Cancel edit' : 'Edit user'}
      </button>
      {isEditing ? (
        <UserEdit {...props} setEditing={setEditing} />
      ) : (
        <UserDisplay {...props} />
      )}
    </div>
  );
}

function UserDisplay({
  name,
  avatar = 'https://via.placeholder.com/100',
  username,
  website,
  email,
  role,
  roleEntities,
}) {
  const {title} = roleEntities[role];
  return (
    <div>
      <p>{name}</p>
      <img src={avatar} alt={name} />
      <p>{username}</p>
      <p>
        <a href={`mailto:${email}`}>{email}</a>
      </p>
      <p>
        <a href={website}>{website}</a>
      </p>
      <p>{title}</p>
    </div>
  );
}

function UserEdit({roleEntities, setEditing, ...restProps}) {
  const initialInputValues = {...restProps};
  const {id, name} = restProps;
  const onSubmit = (dispatch, user) => {
    return dispatch(updateSingleUser(id, user)).then(() => setEditing(false));
  };
  return (
    <UserEditForm
      legendText={`Edit ${name}`}
      buttonText={`Save changes to ${name}`}
      initialInputValues={initialInputValues}
      onSubmit={onSubmit}
      roleEntities={roleEntities}
    />
  );
}
