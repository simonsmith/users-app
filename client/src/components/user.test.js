import React from 'react';
import 'jest-dom/extend-expect';
import {cleanup, render, fireEvent} from '@testing-library/react';
import {User} from './user';
import {configureStore} from 'redux-starter-kit';
import {renderWithStore} from '../test-util/render-with-store';
import {updateSingleUser} from '../store/users';

afterEach(cleanup);
jest.mock('../store/users');

const user = {
  name: 'foo',
  id: '123345',
  avatar: 'avatar.png',
  username: 'fooer',
  website: 'foo.com',
  email: 'me@test.com',
  role: '1',
};
const roleEntities = {
  1: {
    title: 'developer',
  },
};

const renderUser = () => render(<User {...user} roleEntities={roleEntities} />);
const renderUserWithStore = () => {
  const middleware = () => () => () => {
    return Promise.resolve();
  };
  const store = configureStore({
    reducer: state => state,
    middleware: [middleware],
  });
  return renderWithStore(<User {...user} roleEntities={roleEntities} />, store);
};
const clickEditAndGetInputValue = label => {
  const {queryByText, queryByLabelText} = renderUserWithStore();
  fireEvent.click(queryByText('Edit user'));
  return queryByLabelText(label);
};

describe('displaying user', () => {
  test('renders name', () => {
    const {queryByText} = renderUser();
    expect(queryByText('foo')).toBeInTheDocument();
  });

  test('renders username', () => {
    const {queryByText} = renderUser();
    expect(queryByText('fooer')).toBeInTheDocument();
  });

  test('renders website', () => {
    const {queryByText} = renderUser();
    expect(queryByText('foo.com')).toBeInTheDocument();
  });

  test('renders email', () => {
    const {queryByText} = renderUser();
    expect(queryByText('me@test.com')).toBeInTheDocument();
  });

  test('renders role', () => {
    const {queryByText} = renderUser();
    expect(queryByText('developer')).toBeInTheDocument();
  });

  test('renders avatar', () => {
    const {queryByAltText} = renderUser();
    expect(queryByAltText('foo')).toBeInTheDocument();
  });
});

describe('editing a user', () => {
  test('renders an edit button', () => {
    const {queryByText} = renderUser();
    expect(queryByText('Edit user')).toBeInTheDocument();
  });

  test('changes button text when clicking edit', () => {
    const {queryByText} = renderUserWithStore();
    fireEvent.click(queryByText('Edit user'));
    expect(queryByText('Cancel edit')).toBeInTheDocument();
  });

  test('renders name as an input', () => {
    expect(clickEditAndGetInputValue('Name')).toHaveValue('foo');
  });

  test('renders username as an input', () => {
    expect(clickEditAndGetInputValue('Username')).toHaveValue('fooer');
  });

  test('renders website as an input', () => {
    expect(clickEditAndGetInputValue('Website')).toHaveValue('foo.com');
  });

  test('renders email as an input', () => {
    expect(clickEditAndGetInputValue('Email')).toHaveValue('me@test.com');
  });

  test('renders role as a select', () => {
    expect(clickEditAndGetInputValue('Select a role')).toHaveValue('1');
  });
});

describe('submitting the form', () => {
  test('calls the update user action and switches back to user display', async () => {
    updateSingleUser.mockReturnValue({type: 'bar'});
    const {queryByText, findByText} = renderUserWithStore();
    fireEvent.click(queryByText('Edit user'));

    const btn = await findByText('Save changes to foo');
    fireEvent.click(btn);

    const editBtn = await findByText('Edit user');
    const name = queryByText('foo');

    expect(editBtn).toBeInTheDocument();
    expect(name).toBeInTheDocument();
    expect(updateSingleUser.mock.calls).toMatchSnapshot();
  });
});
