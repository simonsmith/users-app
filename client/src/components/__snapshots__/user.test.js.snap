// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`submitting the form calls the update user action and switches back to user display 1`] = `
Array [
  Array [
    "123345",
    Object {
      "avatar": "avatar.png",
      "email": "me@test.com",
      "id": "123345",
      "name": "foo",
      "role": Object {
        "title": "developer",
      },
      "username": "fooer",
      "website": "foo.com",
    },
  ],
]
`;
